### ISSUE TITLE
REST API should support multiple active sessions per user

### ISSUE THREAD

**Issue by tdonohue** [DSpace/DSpace#3257]
**Describe the bug**
Discovered as part of #3249 (original discussion ticket).  Currently, the REST API doesn't support multiple active sessions per user.  For example, while you can login as the same account from multiple browsers / devices, they share the same active session.  Therefore, as soon as a logout occurs, all logged in devices have their session invalidated (logging out the user from all devices).

**To Reproduce**
1. Login in Chrome as dspacedemo+admin
3. Login in Firefox as dspacedemo+admin
4. Logout from Firefox
5. The Chrome session will also be invalidated. You will need to login again to perform any actions requiring authentication.

**Expected behavior**
In the above example, the expectation is that you should still be logged in via Chrome even if you logout on Firefox. These should be treated as separate sessions as they are unrelated, separate "devices". However, we need not support multiple sessions per *browser tab* (as only one session should exist per browser/device).  This will likely require having multiple valid "session salts" for each user account, to allow that same user to be logged in from separate locations.  It also will likely require the creation of a unique identifier (perhaps stored in a browser cookie) to ensure the same session is shared across tabs, but not across different browsers.

While this scenario is not often encountered in Production (as ideally you shouldn't login from several devices at a time), it may be encountered on demo sites (especially during Testathon).  That's why this is currently `low priority`.

**Related work**
Related to #3249 (original discussion issue). In that ticket, this was already estimated as ~3 days of REST Work. Some Angular work (unestimated) is also likely required to manage/create the unique session identifier.

**Comment by saschaszott** [comment]
@tdonohue , @abollini , why do we need to manage the session salt in the database table eperson? Would it not be sufficient  to hold the session salt in-memory (in the user session)? I am not aware of the underlying design decision. The only reason I can image is to support multiple DSpace backend instances ("clustered mode") that are running against the same (shared) database.

**Comment by tdonohue** [comment]
@saschaszott : I believe the reason for the `session_salt` being in the database is that the REST API is "stateless".  In other words, there is no session (at all) on the backend.  All user state information is stored in the JWT, and then the JWT is verified against the `session_salt` in the database.

So, I suspect the proper fix here may require allowing the ability to store multiple `session_salt` values (at once) in the database.  Maybe in a new `eperson_session` table, or similar.  This has another advantage in that (eventually) we may be able to notify the user if they have multiple sessions open, and even allow them to "kill all other sessions" (as some other apps allow you to do these days, like Gmail, etc).

That all said, I have to admit, I haven't dug very deeply on this topic.  What I described above is just an example of how we might achieve this given that we have a stateless backend (with no in-memory session).  There may also be other options/solutions that would work better.

**Comment by saschaszott** [comment]
HI @tdonohue , thank you. I would not propose to make the DSpace backend stateful (or more stateful as it is) but to hold the session_salt in-memory (in the user session). IMHO it is not a discussion about _stateless_ or _stateful_. It is a design decision _where_ we want to store the session_salt within the DSpace backend (in-memory in the app server or on disk in the database layer).

**Comment by saschaszott** [comment]
In our internal discussion we came up with some alternative ideas to solve the problem:

## Idea 1

Add a column `last_request_time` to database table `eperson` (or move this information in the user session; see in-memory vs. on-disk discussion above). This new column holds the timestamp of the last request of a given eperson. The backend takes care that the timestamp is automatically updated on each request (of the given eperson). Note that it is not the same as `last_active` as this column is not updated with _each_ request but only occasionally (when a new JWT token is generated).

The automatic logout in the UI which is triggered in 

https://github.com/DSpace/dspace-angular/blob/main/src/app/shared/idle-modal/idle-modal.component.ts#L53

sends a `/api/authn/logout` POST request. When the backend receives this POST request, it performs an additional check before performing the actual invalidation of the JWT token. If the duration between the current timestamp and the timestamp in `last_request_time` is smaller than idle + grace period (which is currently configured in the UI configuration in `auth.ui.timeUntilIdle` and `auth.ui.idleGracePeriod`) the logout request is **ignored**. The backend signals this fact by returning a certain response code (or response payload) to the UI. The session expiry warning modal dialog in the UI is removed and the idle timer is reset. The eperson is still logged-in (in all tabs / windows).

## Idea 2

If the session expiry warning modal dialog is shown in a tab / browser window, the UI informs the backend via a REST call about this event. Subsequently, either the backend informs all clients (tabs / windows where the current user is logged-in) about this event or each client is regularly asking the backend if such an event occurs. When a client receives the information about the fact that the session expiry warning modal dialog is shown (in another tab / window) it pops up the modal dialog.

This solution guarantees that the session expiry warning modal dialog is shown in each active client (tab / window).

**Comment by tdonohue** [comment]
@saschaszott : Just to clarify what we mean when we say the REST API is built to be "stateless".  We mean we are not storing anything *in memory* on the server.  This allows for the ability to potentially scale DSpace to allow multiple UIs and multiple backends to be spun up when one gets overwhelmed.  So, even if you are "logged in" via Backend 1, your session can swap over to Backend 2 easily, as long as Backend 1 and Backend 2 either use the same database or have a mirrored/replicated database.   

This stateless behavior is more friendly to container based environments, as you potentially can spin up many backend instances when one gets overwhelmed, and your logged in users can use whichever backend is free at any given time (without having to reauthenticate).

So, in this sort of scenario, we cannot store anything "in memory" on the backend, because that means your session is specific to that backend instance.  This is why we currently store the "session_salt" in the database, as this is assumed to be shared (or mirrored) across all backend instances (as all DSpace backends rely on a database).

**Comment by saschaszott** [comment]
I totally agree with you, @tdonohue . If DSpace has to support distributed operation modes (multiple UIs and/or multiple backends / server-applications) than we need to move all state information in the (shared) database layer. In this case, we should consider using Spring Session (https://spring.io/projects/spring-session) which 

>  support clustered sessions without being tied to an application container specific solution

Spring Session

> also provides transparent integration with `HttpSession` and allows replacing the `HttpSession` in an application container (i.e. Tomcat) neutral way, with support for providing session IDs in headers to work with RESTful APIs

**Comment by tdonohue** [comment]
@saschaszott : It is my understanding that Spring Session often stores session information in a Database (or Redis).  So, it's basically doing the same thing we already do.  I'm not against using it, but I want to point out that this aligns very much with our approach of storing `session_salt` in the database.  That "session_salt" is really all the "session" information required to be stored on the backend (at least at this time).

Regardless, my comments above are meant to **describe the current situation and why DSpace 7/8 was designed in this way**.  Those comments are not meant to discount or deprioritize this bug ticket.  I agree this is a bug & it requires a solution.  But, we need to keep in mind **why DSpace works the way it does**, and only significantly change that design if we have **very good reasons** for doing so.
